### PR TITLE
BUG 2071854: Fix unpausing machine config pool too soon

### PR DESCRIPTION
**Pull request by Vincent056**
Fix issues of unpausing machine config pool too soon after applying remediations,
added a check to check if kubeletconfig has been fully rendered into machine config
before unpause affected pool. [1]
([1]https://bugzilla.redhat.com/show_bug.cgi?id=2071854).

### CODE REVIEW

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 1]
```diff
@@ -513,10 +513,18 @@ func (r *ReconcileComplianceSuite) reconcileRemediations(suite *compv1alpha1.Com
 			return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
 		}
 	}
-
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/compliancesuite_controller.go:516]: Unrelated whitespace change?

File: pkg/utils/nodeutils.go [hunk 2]
```diff
@@ -115,6 +121,80 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
...
+		return false, fmt.Errorf("kubeletconfig %s is not subset of %s", currentKCMCName, decodedKC)
+	}
+}
+
+func GetKCFromMC(mc *mcfgv1.MachineConfig, client runtimeclient.Client) (*mcfgv1.KubeletConfig, error) {
+	if mc == nil {
+		return nil, fmt.Errorf("machine config is nil")
+	}
+	if len(mc.GetOwnerReferences()) != 0 {
+		if mc.GetOwnerReferences()[0].Kind == "KubeletConfig" {
```
Comment (rhmdnd) [inline pkg/utils/nodeutils.go:185]: Of this list of references, we only check the first? Or does this mean it's safe to assume there is only ever one element in the list?

File: pkg/utils/compare_json.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	table.Render()
+}
+
+// JSONIsSubset checks if a is a subset json of b
+func JSONIsSubset(a, b []byte) (bool, *JSONDiff, error) {
+	return jsonIsSubsetR(a, b, nil, nil)
+}
+
+func jsonIsSubsetR(a, b []byte, diff *JSONDiff, prefix interface{}) (bool, *JSONDiff, error) {
```
Comment (rhmdnd) [inline pkg/utils/compare_json.go:44]: There are a lot of cases in here. Can we add some unit tests or is it covered in an integration test?

File: pkg/utils/nodeutils.go [hunk 4]
```diff
@@ -115,6 +121,80 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
 	return true, currentKCMC, nil
 }
 
+func AreKubeletConfigsRendered(pool *mcfgv1.MachineConfigPool, client runtimeclient.Client) (bool, error) {
```
Comment (rhmdnd) [inline pkg/utils/nodeutils.go:124]: Similar comment as above about testing.

File: CHANGELOG.md [hunk 5]
```diff
@@ -31,6 +31,22 @@ Versioning](https://semver.org/spec/v2.0.0.html).
...
+
+### Enhancements
+
+- 
+
+### Fixes
+
+- Fix issues of unpausing machine config pool too soon after apply remediations,
+  added a check to check if kubelet has been fully rendered into machine config
+  before unpause affected pool. [1]
```
Comment (rhmdnd) [inline CHANGELOG.md:44]: nit: I'd try and include the outcome here if possible (helps justify why we made the change).

For example:

The compliance-operator now waits until the kubelet configuration is fully rendered before unpausing the pool to avoid kubelets getting stuck in a `NotReady` state.

File: pkg/utils/nodeutils.go (hunk 2 above)
Comment (Vincent056) [reply pkg/utils/nodeutils.go:185]: It is safe to assume there is only ever one element in the list as only one KC corresponds to one MC

File: pkg/utils/compare_json.go (hunk 3 above)
Comment (Vincent056) [reply pkg/utils/compare_json.go:44]: @rhmdnd I just added integration test here: https://github.com/openshift/compliance-operator/pull/814/files#diff-0a9f36cc9b36d2e583d93013174fdc0bfedafad46ef1e2075effcbf8bbb87e8dR604

I am not sure if we could add an e2e test  for this one since it will be hard to mimic the delay for rendering KubeletConfig, but I think having this test should be enough. This fix will greatly improve the stability of our compliance operator.

File: pkg/controller/complianceremediation/complianceremediation_controller.go [hunk 6]
```diff
@@ -240,6 +240,7 @@ func (r *ReconcileComplianceRemediation) reconcileRemediation(instance *compv1al
 	}
 
 	if instance.Spec.Apply {
+		fmt.Println("something")
```
Comment (jhrozek) [inline pkg/controller/complianceremediation/complianceremediation_controller.go:243]: this leftover should be removed

File: pkg/utils/nodeutils.go [hunk 7]
```diff
@@ -115,6 +121,79 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
...
+	isUsingKC, currentKCMCName, err := IsMcfgPoolUsingKC(pool)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if pool %s is using a custom kubelet config: %w", pool.Name, err)
+	}
+	if !isUsingKC || currentKCMCName == "" {
+		return true, nil
+	}
+	// if the pool is using a custom kubelet config, check if the kubelet config is rendered
+	kcmcfg := &mcfgv1.MachineConfig{}
+	err = client.Get(context.TODO(), types.NamespacedName{Name: currentKCMCName, Namespace: pool.Namespace}, kcmcfg)
```
Comment (jhrozek) [inline pkg/utils/nodeutils.go:135]: This is a nit, but a pool is not namespaced, but cluster-wide, so you can just remove the `Namespace` parameter and not use the `pool.Namespace` attribute, it will always be an empty string anyway.

File: pkg/utils/nodeutils.go [hunk 8]
```diff
@@ -115,6 +121,79 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
...
+
+	encodedKC, err := jsonpath.Get("storage.files[0].contents.source", obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get encoded kubelet config from machine config %s: %w", currentKCMCName, err)
+	}
+	encodedKCStr := encodedKC.(string)
+	if encodedKCStr == "" {
+		return false, fmt.Errorf("encoded kubeletconfig %s is empty", currentKCMCName)
+	}
+	mcPayloadPrefix := `data:text/plain,`
```
Comment (jhrozek) [inline pkg/utils/nodeutils.go:159]: can you make this a const?

File: pkg/utils/nodeutils.go [hunk 9]
```diff
@@ -115,6 +121,79 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
...
+	encodedKCStr = strings.TrimPrefix(encodedKCStr, mcPayloadPrefix)
```
Comment (jhrozek) [inline pkg/utils/nodeutils.go:160]: If we are expecting the prefix, shouldn't we be more defensive and error out if the prefix is not there?

File: pkg/utils/nodeutils.go [hunk 10]
```diff
@@ -115,6 +121,79 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
...
+	}
+
+	isSubset, diff, err := JSONIsSubset(kc.Spec.KubeletConfig.Raw, []byte(decodedKC))
+	if err != nil {
+		return false, fmt.Errorf("failed to check if kubeletconfig %s is subset of %s: %w", currentKCMCName, decodedKC, err)
+	}
+	if isSubset {
+		return true, nil
+	} else {
+		PrintJSONDiff(diff)
```
Comment (jhrozek) [inline pkg/utils/nodeutils.go:174]: hmm, on one hand I like printing the diff, on the other hand shouldn't we go through the logger interface rather than printing the diff directly to stdout? If you don't like to add logging to this function, returning the diff up or writing it into a io.Writer or smh would both work better IMO than an utility function printing to stdout..

**Review by jhrozek — COMMENTED** [review]
See a couple of nits inline, mostly looks good!

File: pkg/utils/nodeutils.go (hunk 7 above)
Comment (Vincent056) [reply pkg/utils/nodeutils.go:135]: that's a good catch!

**Comment by jhrozek** [comment]
You need to rebase atop origin/master to get rid of the merge conflict or else CI won't run..

**Review by jhrozek — APPROVED** [review]
/lgtm
thank you, all my changes have been addressed
